src/resty/gcp/init.lua

```diff
@@ -51,50 +51,116 @@ FindApis = function(apiClass, methods, curr)
     return methods
 end
 
+local function template_expansion(str, params)
+    return (string.gsub(str, "{(.-)}", function (key)
+        return ngx.escape_uri(params[key])
+    end))
+end
+
+local function build_request(accesstoken, apiDetail, baseUrl, params, requestBody, opt)
+    local media = opt and opt.media
+    if media and not apiDetail.supportsMediaDownload then
+        error("API does not supported media download")
+    end
+
+    local mediaUpload = apiDetail.supportsMediaUpload
+
+    local req = {
+        method = apiDetail.httpMethod,
+        headers = {
+            ["Authorization"] = "Bearer " .. accesstoken.token
+        },
+        body = requestBody,
+        ssl_verify = false
+    }
+
+    -- simple APIs
+    if apiDetail.flatPath then
+        local path = template_expansion(apiDetail.flatPath, params)
+        return path, req
+    end
+
+    local path_template
+    if mediaUpload then
+        path_template = "https://storage.googleapis.com" .. apiDetail.mediaUpload.protocols.simple.path
+    else
+        path_template = baseUrl .. apiDetail.path
+    end
```

Comment:
Man, the GCP API is _weird_.

Reply:
@flrgh I agree. They use 2 different styles to record URLs in the definition for API support or not support media download.